shadows/framework/src/main/java/org/robolectric/shadows/ShadowBiometricManager.java

```diff
@@ -37,21 +42,50 @@ public class ShadowBiometricManager {
   protected int canAuthenticate() {
     if (RuntimeEnvironment.getApiLevel() >= R) {
       return reflector(BiometricManagerReflector.class, realBiometricManager).canAuthenticate();
-    } else if (biometricServiceConnected) {
-      return BIOMETRIC_SUCCESS;
     } else {
-      boolean hasBiomatrics =
-          ReflectionHelpers.callStaticMethod(
-              BiometricManager.class,
-              "hasBiometrics",
-              ClassParameter.from(
-                  Context.class, RuntimeEnvironment.getApplication().getApplicationContext()));
-      if (!hasBiomatrics) {
-        return BIOMETRIC_ERROR_NO_HARDWARE;
-      } else {
-        return BIOMETRIC_ERROR_HW_UNAVAILABLE;
-      }
+      return getBiometricResult();
+    }
+  }
+
+  @RequiresPermission(USE_BIOMETRIC)
+  @Implementation
+  protected int canAuthenticate(int authenticators) {
+    return getBiometricResult();
+  }
+
+  @RequiresPermission(USE_BIOMETRIC)
+  @Implementation(minSdk = R)
+  protected int canAuthenticate(int userId, int authenticators) {
+    return getBiometricResult();
+  }
+
+  private int getBiometricResult() {
```

Comment:
Maybe this method accepts `int userId` and `authenticators`, and all `canAuthenticate` call this method as current logic. For `canAuthenticate()` and `canAuthenticate(int authenticators)`, we can pass default value for `userId` or `authenticators` or both of them as AOSP does. IMO, the name `canAuthenticateInternal` is better than `getBiometricResult()`.

Reply:
sure 👌